iOS/UZTickets/src/scan/UZTScanViewController.m

```diff
@@ -0,0 +1,180 @@
+//
+//  UZTScanViewController.m
+//  UZTickets
+//
+//  Created by nekoi on 6/1/15.
+//  Copyright (c) 2015 DAloG ltd. All rights reserved.
+//
+
+#import "UZTScanViewController.h"
+#import "UZTScanView.h"
+
+@interface UZTScanViewController ()<AVCaptureMetadataOutputObjectsDelegate>
+
+@property UZTScanView* view;
+@property (strong) AVCaptureMetadataOutput* output;
+@property (strong) AVCaptureSession* session;
+@property (strong) dispatch_queue_t dispatchQueue;
+
+@end
+
+@implementation UZTScanViewController
+@dynamic view;
+
++ (BOOL)canScan
+{
+    NSArray* devices = [AVCaptureDevice devicesWithMediaType:AVMediaTypeVideo];
+    if (devices.count == 0) return NO;
+    
+    AVCaptureDevice* device;
+    
+    for (AVCaptureDevice* cameraDevice in devices)
+    {
+        if (cameraDevice.position == AVCaptureDevicePositionBack)
+        {
+            device = cameraDevice;
+            break;
+        }
+    }
+    
+    if (device == nil) return NO;
+    
+    NSError* inputError;
+    AVCaptureDeviceInput* input = [AVCaptureDeviceInput deviceInputWithDevice:device error:&inputError];
+    NSAssert(inputError == nil, @"An error occurred while try get AVCaptureDeviceInput: %@", inputError);
+    if (inputError != nil) return NO;
+    
+    AVCaptureSession* session = [AVCaptureSession new];
+    if (![session canAddInput:input]) return NO;
+    
+    return YES;
+}
+
+- (void)setupLayerSessionWithAvailableObjects:(NSArray*)availableObjectTypes
+{
+    [self.viewModel didStartInitializingScanView];
```

Comment:
lets solve problems by priority. We can add some code always. I’ll drop this.

> 1 июня 2015, в 16:48, Kostya <notifications@github.com> написал(а):
> 
> In iOS/UZTickets/src/scan/UZTScanViewController.m <https://github.com/ua-apps/uz-tickets/pull/5#discussion_r31426370>:
> 
> > +    if (device == nil) return NO;
> > +    
> > +    NSError* inputError;
> > +    AVCaptureDeviceInput* input = [AVCaptureDeviceInput deviceInputWithDevice:device error:&inputError];
> > +    NSAssert(inputError == nil, @"An error occurred while try get AVCaptureDeviceInput: %@", inputError);
> > +    if (inputError != nil) return NO;
> > +    
> > +    AVCaptureSession* session = [AVCaptureSession new];
> > +    if (![session canAddInput:input]) return NO;
> > +    
> > +    return YES;
> > +}
> > +
> > +- (void)setupLayerSessionWithAvailableObjects:(NSArray*)availableObjectTypes
> > +{
> > +    [self.viewModel didStartInitializingScanView];
> As You've mentioned AV is havy to initialize, therefore it will be needed that will be needed to sometimes show activity indicator while system set up all the things for capturing and representing data.
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/ua-apps/uz-tickets/pull/5/files#r31426370>.
>